editor/plugins/path_3d_editor_plugin.cpp

```diff
@@ -303,33 +303,45 @@ void Path3DGizmo::redraw() {
 		}
 
 		const Transform3D *r = frames.ptr();
+
 		Vector<Vector3> _collision_segments;
+		_collision_segments.resize((sample_count - 1) * 2);
+		Vector3 *_collisions_ptr = _collision_segments.ptrw();
+
 		Vector<Vector3> bones;
+		bones.resize(sample_count * 4);
+		Vector3 *bones_ptr = bones.ptrw();
+
 		Vector<Vector3> ribbon;
+		ribbon.resize(sample_count);
+		Vector3 *ribbon_ptr = ribbon.ptrw();
+
 		for (int i = 0; i < sample_count; i++) {
 			const Vector3 p1 = r[i].origin;
 			const Vector3 side = r[i].basis.get_column(0);
 			const Vector3 up = r[i].basis.get_column(1);
 			const Vector3 forward = r[i].basis.get_column(2);
 
 			// Collision segments.
-			if (i != sample_count) {
+			if (i != sample_count - 1) {
 				const Vector3 p2 = r[i + 1].origin;
-				_collision_segments.push_back(p1);
-				_collision_segments.push_back(p2);
+				_collisions_ptr[(i * 2)] = p1;
+				_collisions_ptr[(i * 2) + 1] = p2;
 			}
 
 			// Path3D as a ribbon.
-			ribbon.push_back(p1);
+			ribbon_ptr[i] = p1;
 
 			// Fish Bone.
 			const Vector3 p_left = p1 + (side + forward - up * 0.3) * 0.06;
 			const Vector3 p_right = p1 + (-side + forward - up * 0.3) * 0.06;
-			bones.push_back(p1);
-			bones.push_back(p_left);
 
-			bones.push_back(p1);
-			bones.push_back(p_right);
+			const int bone_idx = i * 4;
+
+			bones_ptr[bone_idx] = p1;
+			bones_ptr[bone_idx + 1] = p_left;
+			bones_ptr[bone_idx + 2] = p1;
+			bones_ptr[bone_idx + 3] = p_right;
 		}
 
 		add_collision_segments(_collision_segments);
```

Comment:
Just because I am not familiar with this code, can you tell me what the practical result of having fewer ``_collision_segments`` is? Other than not reading outside the bounds of the ``frames`` vector

Reply:
The amount of segments should be equal to the number of samples, except the final one. Each segment has a start & end point, and the start point of every segment after the first one is the same as the previous one's end point. 

Basically, this is how the collision segments should work:
![segments](https://github.com/godotengine/godot/assets/53150244/9f391feb-3b49-4567-bd3d-0e44e4cbf0ca)

Previously, the if statement would've always gotten triggered, and as a result an invalid segment would get added to the array, like this:
![segments_bad](https://github.com/godotengine/godot/assets/53150244/be1d52be-0e8f-4b2e-9f48-ac70623303ee)

Reply:
Thanks for the explanation!